python_modules/dagster/dagster/_core/definitions/definitions_class.py

```diff
@@ -0,0 +1,92 @@
+from typing import Any, Dict, Iterable, Mapping, Optional, Union
+
+from dagster._annotations import experimental
+from dagster._core.execution.with_resources import with_resources
+
+from .assets import AssetsDefinition, SourceAsset
+from .cacheable_assets import CacheableAssetsDefinition
+from .decorators import repository
+from .job_definition import JobDefinition
+from .repository_definition import (
+    SINGLETON_REPOSITORY_NAME,
+    PendingRepositoryDefinition,
+    RepositoryDefinition,
+)
+from .resource_definition import ResourceDefinition
+from .schedule_definition import ScheduleDefinition
+from .sensor_definition import SensorDefinition
+
+
+@experimental
+class Definitions:
+    def __init__(
+        self,
+        assets: Optional[
+            Iterable[Union[AssetsDefinition, SourceAsset, CacheableAssetsDefinition]]
+        ] = None,
+        schedules: Optional[Iterable[ScheduleDefinition]] = None,
+        sensors: Optional[Iterable[SensorDefinition]] = None,
+        jobs: Optional[Iterable[JobDefinition]] = None,
+        resources: Optional[Mapping[str, Any]] = None,
+    ):
+        """
+        Example usage:
+
+        defs = Definitions(
+            assets=[asset_one, asset_two],
+            schedules=[a_schedule],
+            sensors=[a_sensor],
+            jobs=[a_job],
+            resources={
+                "a_resource": some_resource,
+            }
+        )
+
+        Create a set of definitions explicitly available and loadable by dagster tools.
+
+        Dagster separates user-defined code from system tools such the web server and
+        the daemon. Rather than loading code directly into process, a tool such as the
+        webserver interacts with user-defined code over a serialization boundary.
+
+        These tools must be able to locate and load this code when they start. Via CLI
+        arguments or config, they specify a python module to inspect.
+
+        A python module is loadable by dagster tools if it there is a top level variable
+        named `defs` that is an instance of Definitions.
+
+        Definitions provides a few conveniences for dealing with resources that do not apply to
+        vanilla dagster definitions:
+
+        (1) It takes a dictionary of top-level resources which are automatically bound
+        (via with_resources) to any asset passed to it. If you need to apply different
+        resources to different assets, use legacy @repository and use with_resources as before.
+
+        (2) The resources dictionary takes raw python objects, not just resource definitions.
+        """
+
```

Comment:
should we type-check immediately to avoid harder-to understand errors from inside created_repo (if you pass in a bad object type)?